# Creditors reliability research.

### Project description 

<u>Customer</u> — credit department of the bank. Input data from bank - statistics on the solvency of customers for the previous period.

<u>The purpose of the study</u> is to check whether the fact of repaying the loan on time affects:
1. Marital status of the borrower.
2. Number of children of the borrower.
3. Income of the borrower.
4. Purpose of the loan.
 
The results of research will be taken into account when building a credit scoring model - a special system that assesses the ability of a potential borrower to repay a loan to a bank.

Research sequence:
 1. Data review.
 2. Data preprocessing.
 3. Hypothesis testing.

### Data review

In [1]:
import pandas as pd 
try:
    data = pd.read_csv('~/Documents/programming/data_preprocessing/data.csv')  # Локальный путь
except:
    data = pd.read_csv('/data.csv')  # Серверный путь
data.info()
data.head(30)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   
20         1   -1311.604166         36              среднее             1   
21         1    -253.685166         33              среднее             1   
22         1   -1766.644138         24              среднее             1   
23         0    -272.981385         21               высшее             0   
24         1  338551.952911         57              среднее             1   
25         0  363548.489348         67              среднее             1   
26         0            NaN         41              среднее             1   
27         0    -529.191635         28               высшее             0   
28         1    -717.274324         26               высшее             0   
29         0            NaN         63              среднее             1   

            family_status  family_status_id gender  income_type  debt  \
0         женат / замужем                 0      F    сотрудник     0   
1         женат / замужем                 0      F    сотрудник     0   
2         женат / замужем                 0      M    сотрудник     0   
3         женат / замужем                 0      M    сотрудник     0   
4        гражданский брак                 1      F    пенсионер     0   
5        гражданский брак                 1      M    компаньон     0   
6         женат / замужем                 0      F    компаньон     0   
7         женат / замужем                 0      M    сотрудник     0   
8        гражданский брак                 1      F    сотрудник     0   
9         женат / замужем                 0      M    сотрудник     0   
10        женат / замужем                 0      M    компаньон     0   
11        женат / замужем                 0      F    сотрудник     0   
12       гражданский брак                 1      M    пенсионер     0   
13        женат / замужем                 0      F    сотрудник     0   
14       гражданский брак                 1      F    компаньон     1   
15        женат / замужем                 0      F    сотрудник     0   
16        женат / замужем                 0      F    сотрудник     0   
17       гражданский брак                 1      M    сотрудник     0   
18         вдовец / вдова                 2      F    пенсионер     0   
19              в разводе                 3      F    сотрудник     0   
20        женат / замужем                 0      M    сотрудник     0   
21       г

In [2]:
data.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

There are 12 columns in the dataset, there are no problems with column names, types are also more or less normal at first glance, numeric - float or int, text - type object. In numeric, it will be necessary to convert some of the data to int, since, for example, experience cannot be fractional. In the values themselves, problems are already visible, for example, a negative seniority value, different case of spelling in the education column, as well as NaN values in some columns.

By the number of values, you can see that NaN values are in the days_employed and total_income columns and there are 2174 of them. This is 10% of all data. A rather high percentage of gaps and ignoring these data can affect the results of our study, so we will replace the missing values with medians.

The reasons for the appearance of omissions are most likely technical, since this is data from the bank for the previous period and it is impossible to consider a loan application without indicating the length of service and income, these are quite important indicators when considering.

We replace the missing values with medians, since both the experience column and the income column have values that are very different from the majority and the average value can greatly distort the picture.

### Data preprocessing

Let's translate the negative values of experience into positive ones.

In [3]:
data['days_employed'] = data['days_employed'].abs() 

Let's calculate the median values of seniority and income, to fill in the gaps, depending on the type of income, as this allows us to assume that the indicators for the same types are approximately similar. For example, the experience and income of different students plus or minus should be similar.

In order not to make a lot of movements in the calculations, we will determine what unique values of the income type are and using a loop, we will determine the median value and replace the gaps in a certain group by income type with it.

In [4]:
income_types = data['income_type'].unique()
print('All types of income:', income_types)
print()
for i in income_types:
    income_set = data['income_type'] == i
    income_median = data[income_set]['total_income'].median()
    days_median = data[income_set]['days_employed'].median()
    print('Income and seniority median for type', i, income_median, 'and', days_median)
    data.loc[income_set, 'total_income'] = data.loc[income_set, 'total_income'].fillna(income_median)
    data.loc[income_set, 'days_employed'] = data.loc[income_set, 'days_employed'].fillna(days_median)

All types of income: ['сотрудник' 'пенсионер' 'компаньон' 'госслужащий' 'безработный'
 'предприниматель' 'студент' 'в декрете']

Income and seniority median for type сотрудник 142594.39684740017 and 1574.2028211070854
Income and seniority median for type пенсионер 118514.48641164352 and 365213.3062657312
Income and seniority median for type компаньон 172357.95096577113 and 1547.3822226779334
Income and seniority median for type госслужащий 150447.9352830068 and 2689.3683533043886
Income and seniority median for type безработный 131339.7516762103 and 366413.65274420456
Income and seniority median for type предприниматель 499163.1449470857 and 520.8480834953765
Income and seniority median for type студент 98201.62531401133 and 578.7515535382181
Income and seniority median for type в декрете 53829.13072905995 and 3296.7599620220594


Also, after the replacement, we will check for the correct filling of gaps based on rows with index 12 and 26, where there were NaN values in the columns total_income and days_employed

In [5]:
print('check for total_income')
print(data.loc[12, 'income_type'], data.loc[12, 'total_income'])
print(data.loc[26, 'income_type'], data.loc[26, 'total_income'])
print()
print('check for days_employed')
print(data.loc[12, 'income_type'], data.loc[12, 'days_employed'])
print(data.loc[26, 'income_type'], data.loc[26, 'days_employed'])

check for total_income
пенсионер 118514.48641164352
госслужащий 150447.9352830068

check for days_employed
пенсионер 365213.3062657312
госслужащий 2689.3683533043886


Now we can be sure that the replacement happened correctly, the value was substituted depending on the type of income.

Also using the info() method, make sure that there are no more empty values.

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


### Checking data for anomalies and corrections.

We check the remaining columns for anomalies.
1) As a result of the check, it was found that for the category "pensioners" the length of service in days is always higher than 328000, which is more than 900 years, this is a clear problem. Most likely, for some technical reason, the seniority of pensioners is indicated not in days, but in hours, but for the purposes of this study, this error does not matter, since we take the median, not the average, to fill in the missing values in the days_employed column.

2) Also in the dob_years column, which means the age of the borrower is 101, the value is "0" years. Most likely this is a technical error, as there are too many repetitions of the error. I would pass this information on to the technicians for investigation. For the conclusions of this study, this error is not important, but for practice I will replace 0 with the median value for the column.

3) But in the column the number of children is the value -1 and 20. In my opinion, these are the values 1 and 2, respectively. Since such values are not single, they should be reported to technical specialists. And for our study, these data are very important, so we will replace these anomalous values with real ones.

In [7]:
data['dob_years'] = data['dob_years'].replace(0, data['dob_years'].median())

In [8]:
data['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

In [9]:
data['children'] = data['children'].replace(-1, 1)
data['children'] = data['children'].replace(20, 2)
data['children'].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

### Changing data types.

Let's replace the real values in the columns of experience and income with integer ones, since these indicators cannot be fractional. We will use the astype() method. After executing the method, let's look at the changes

In [10]:
data['days_employed'] = data['days_employed'].astype('int')
data['total_income'] = data['total_income'].astype('int')
data.head()

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   Среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

### Removing duplicates.

During initial data examination, it was obvious that in the education column there are options for writing the same value using different case, we will check this using the .value_counts() method 

In [11]:
data['education'].value_counts()

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64

There are indeed many repetitions of the same value using different case, we will eliminate this using the lower() method

In [12]:
data['education'] = data['education'].str.lower()
data['education'].value_counts()

среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

Now there are only 5 unique types in the education column. Let's check the categorical columns family_status, income_type, purpose, gender in the same way, pre-processing some with the lower() method for ease of perception.

In [13]:
data['family_status'] = data['family_status'].str.lower()
data['income_type'] = data['income_type'].str.lower()
data['purpose'] = data['purpose'].str.lower()
data['family_status'].value_counts()

женат / замужем          12380
гражданский брак          4177
не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

In [14]:
data['income_type'].value_counts()

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
студент                1
в декрете              1
Name: income_type, dtype: int64

In [15]:
data['purpose'].value_counts()

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
покупка жилья для сдачи                   653
операции с жильем                         653
операции с коммерческой недвижимостью     651
покупка жилья                             647
жилье                                     647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
строительство недвижимости                620
покупка своего жилья                      620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

In [16]:
data['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

There was one entry in the gender column that looks like a random error with the value xna, this one value does not play a special role, since it is a negligible percentage in relation to the main data, so it can either be deleted or replaced with any of the correct values. I choose to replace xna with m.

In [17]:
data['gender'] = data['gender'].replace('XNA', 'M')
data['gender'].value_counts()

F    14236
M     7289
Name: gender, dtype: int64

Check the data for obvious duplicates

In [18]:
data.duplicated().sum()

72

There are 72 obvious duplicates in the dataset, let's remove them and check this.

In [19]:
data = data.drop_duplicates().reset_index(drop=True)
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21453 non-null  int64 
 1   days_employed     21453 non-null  int64 
 2   dob_years         21453 non-null  int64 
 3   education         21453 non-null  object
 4   education_id      21453 non-null  int64 
 5   family_status     21453 non-null  object
 6   family_status_id  21453 non-null  int64 
 7   gender            21453 non-null  object
 8   income_type       21453 non-null  object
 9   debt              21453 non-null  int64 
 10  total_income      21453 non-null  int64 
 11  purpose           21453 non-null  object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


### Formation of additional dataframes of dictionaries, decomposition of the original dataframe.

Let's create two new dataframes with columns: education_id and education - in the first one; family_status_id and family_status - in the second, in order to unload our dataframe from text, we will remove the education and family_status columns from the original dataframe, leaving only their identifiers, check whether it worked by displaying the first five rows of the dataframe.

In [20]:
education_dict = data[['education_id', 'education']]
education_dict = education_dict.drop_duplicates().reset_index(drop=True)
education_dict

education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень

In [21]:
family_status_dict = data[['family_status_id', 'family_status']]
family_status_dict = family_status_dict.drop_duplicates().reset_index(drop=True)
family_status_dict

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  не женат / не замужем

In [22]:
data.drop(columns = ['family_status', 'education'],axis = 1, inplace=True)
data.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0         340266         53             1                 1      F   

  income_type  debt  total_income                     purpose  
0   сотрудник     0        253875               покупка жилья  
1   сотрудник     0        112080     приобретение автомобиля  
2   сотрудник     0        145885               покупка жилья  
3   сотрудник     0        267628  дополнительное образование  
4   пенсионер     0        158616             сыграть свадьбу

### Income categorization.

Let's divide clients into categories depending on their income using the income_category function, display for clarity using the value_counts method clients of all categories, at the same time check whether representatives of all categories are present in the dataframe.

In [23]:
def income_category(income):
    if income <= 30000:
        return 'E'
    if income <= 50000:
        return 'D'
    if income <= 200000:
        return 'C'
    if income <= 1000000:
        return 'B'
    return 'A'

data['total_income_category'] = data['total_income'].apply(income_category)
data['total_income_category'].value_counts()

C    16014
B     5042
D      350
A       25
E       22
Name: total_income_category, dtype: int64

### Loan purposes categorization.

Dividing customers into categories depending on the purpose of the loan. Also, after the breakdown, let's see how many representatives of different categories. It turned out that the most common purpose of the loan is real estate transactions.

In [24]:
def purpose_category(purpose):
    if 'свадьб' in purpose:
        return 'проведение свадьбы'
    if 'образовани' in purpose:
        return 'получение образования'
    if 'автомобил' in purpose:
        return 'операции с автомобилем'
    return 'операции с недвижимостью'

data['purpose_category'] = data['purpose'].apply(purpose_category)
data['purpose_category'].value_counts()

операции с недвижимостью    10811
операции с автомобилем       4306
получение образования        4013
проведение свадьбы           2323
Name: purpose_category, dtype: int64

### Answers.

In order to understand whether the number of children, marital status, income level, and the purpose of the loan affect its repayment on time, we will build summary tables in which we calculate the share of borrowers in each category who had debt to repay loans (debt = 1).

Let's start the study in groups divided by the number of children:

In [25]:
data_pivot = data.pivot_table(index=['children'],values = 'debt', aggfunc='mean')
data_pivot['group_amount'] = data['children'].value_counts()
data_pivot.sort_values(by='debt', ascending=True).head(10)

debt  group_amount
children                        
5         0.000000             9
0         0.075444         14090
3         0.081818           330
1         0.091658          4855
2         0.094925          2128
4         0.097561            41

As can be seen from the table, among clients without children, the proportion of borrowers who had debts to repay loans is actually lower than among other groups. For a group with 3, 4 and 5 children, the sample is too small to draw unambiguous conclusions. The most representative are the samples with 0, 1 and 2 children. If we analyze only them, then we can assume that borrowers without children are a little more reliable, but the difference is small, only 1.5 - 2%.

Let's calculate the percentage of clients who had loan repayment arrears to the total number of clients by groups by marital status

In [26]:
data_pivot = data.pivot_table(index=['family_status_id'],values = 'debt', aggfunc='mean')
data_pivot['group_amount'] = data['family_status_id'].value_counts()
data_pivot.sort_values(by='debt', ascending=True).head(10)

debt  group_amount
family_status_id                        
2                 0.065693           959
3                 0.071130          1195
0                 0.075452         12339
1                 0.093494          4150
4                 0.097509          2810

Among the categories of married/married or former clients, the proportion of people who had loan repayment arrears is lower than among those who have never been married/married. On average, the difference between these groups is about 2%, the deviation is not large enough to seriously consider it.

Let us calculate the share of clients who had debts to repay loans to the total number of clients by groups by income category.

In [27]:
data_pivot = data.pivot_table(index=['total_income_category'],values = 'debt', aggfunc='mean')
data_pivot['group_amount'] = data['total_income_category'].value_counts()
data_pivot.sort_values(by='debt', ascending=True).head(10)

debt  group_amount
total_income_category                        
D                      0.060000           350
B                      0.070607          5042
A                      0.080000            25
C                      0.084926         16014
E                      0.090909            22

The smallest percentage of clients who had loan repayment arrears is in the group of people with an income of 30 to 50 thousand,
Among millionaires, there are problems with repayment (cat. A), and among borrowers with an income of less than 30 thousand (cat. E). But the sample for categories A and E is too small to seriously consider their results. For category D, the sample is larger, but still not very representative - only 350 rows!

You can seriously consider the results for categories B and C, since the sample size allows this. The difference between them is 1.5%, which is an extremely modest deviation.


Let us calculate the share of clients who had debts to repay loans to the total number of clients by groups according to the purpose of the loan.

In [28]:
data_pivot = data.pivot_table(index=['purpose_category'],values = 'debt', aggfunc='mean')
data_pivot['group_amount'] = data['purpose_category'].value_counts()
data_pivot.sort_values(by='debt', ascending=True).head(10)

debt  group_amount
purpose_category                                
операции с недвижимостью  0.072334         10811
проведение свадьбы        0.080069          2323
получение образования     0.092200          4013
операции с автомобилем    0.093590          4306

As we can see, clients who have a goal of buying real estate and want to have a wedding on credit are more trustworthy than clients who buy a car and want to study on credit.

## Общий вывод:

In the study, I tested 4 questions.
Does the fact of loan repayment on time affect:
1. Marital status of the borrower.
2. Number of children of the borrower.
3. Income of the borrower
4. Purpose of the loan

The following conclusions were made:

1. Marital status does not greatly affect the fact of loan repayment on time, but a borrower who is married or has been married is a little more reliable than those who were not.

2. According to the calculations, it may seem that borrowers with 5 children are ideal, but the sample is so small that it is very unrepresentative and it is wrong to base it on the result. On average, for the rest of the samples, the difference is small. Borrowers without children are only 1.5-2% more reliable than borrowers with children.

3. In terms of the borrower's income, only the results in categories B and C can be seriously considered, since the sample size allows this. The remaining samples are unrepresentatively small. The difference between B and C is 1.5%, which is an extremely modest deviation.

4. According to the purpose of the loan, borrowers who want to buy real estate and have a wedding are more trustworthy than customers who buy a car and want to study on credit. Perhaps this is due to the fact that real estate is usually bought and married by people who have decided and taken place, and education and car loans are more often the work of younger people. But again, the difference between the sample shares is small, about 2%, to base the scoring on.

In general, we can say that none of the factors considered strongly affects the fact of repaying the loan on time. The difference in percentage between groups is no more than 2 percent on average, despite the fact that the sizes of some samples are not representative.

Also, technical specialists would like to convey that there are some problems with the data, such as a rather large number of empty values of seniority and income, the use of a different register in the contents of the education column, numerous errors in the number of children column, for the category "retirees" the length of service in days is always above 328,000, which is over 900 years old. Judging by the number of these errors, this is not a human factor, but a system error.